File: problems/162_find-peak-element/main.cpp
#include <iostream>
#include <utility>
#include <vector>

class Solution
{
public:
    int findPeakElement(std::vector<int> &nums)
    {
        int left = 0, right = nums.size() - 1;
        while (left < right)
        {
            int mid = left + (right - left) / 2;
            if (nums[mid] < nums[mid + 1])
                left = mid + 1;
            else
                right = mid;
        }
        return left;
    }
};

int main()
{
    Solution solution;
    std::vector<std::pair<std::vector<int>, int>> samples = {
        {{1, 2, 3, 1}, 2},
        {{1, 2, 1, 3, 5, 6, 4}, 5},
    };

    for (const auto &[nums, expected] : samples)
    {
        auto input = nums;
        int result = solution.findPeakElement(input);

        std::cout << "input: nums = [";
        for (size_t i = 0; i < nums.size(); ++i)
        {
            std::cout << nums[i] << (i + 1 < nums.size() ? "," : "");
        }
        std::cout << "]"
                  << ", placeholder result: " << result
                  << ", expected: " << expected << '\n';
    }

    return 0;
}

File: problems/162_find-peak-element/main.rs
struct Solution;

impl Solution {
    pub fn find_peak_element(nums: Vec<i32>) -> i32 {
        let mut left = 0;
        let mut right = nums.len() as i32 - 1;
        while left < right {
            let mid = left + (right - left) / 2;
            if nums[mid as usize] < nums[(mid + 1) as usize] {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        left
    }
}

fn main() {
    let samples: Vec<(Vec<i32>, i32)> = vec![(vec![1, 2, 3, 1], 2), (vec![1, 2, 1, 3, 5, 6, 4], 5)];

    for (nums, expected) in samples {
        let result = Solution::find_peak_element(nums.clone());
        println!(
            "input: nums = {:?}, placeholder result: {}, expected: {}",
            nums, result, expected
        );
    }
}
